### PR TITLE
[Feature] Improve API DRF Authentication [OSF-7434]

### PR DESCRIPTION
**Pull request by cslzchen**
### Purpose

Branded Preprints Phase 2 relies on CORS making API calls with OSF cookie for authentication. This ticket revisits `api/base/authentication/drf.py` in the following aspects:

- verify cookie based authentication for API
- update user status check
- use native Django model and query
- update comments and doc string

### Ticket

https://openscience.atlassian.net/browse/OSF-7434

### CODE REVIEW

File: api/base/authentication/drf.py [hunk 1]
```diff
@@ -1,88 +1,166 @@
...
 
 def check_user(user):
-    """Checks that the user is neither unconfirmed nor disabled.
-
-    :param User user:
-    :raises DeactivatedAccountError:
-    :raise UnconfirmedAccountError:
     """
+    Verify users' status.
+
```
Comment (cslzchen) [inline api/base/authentication/drf.py]: A comprehensive user status check based on six properties in `OSFUser`.

File: api/base/authentication/drf.py [hunk 2]
```diff
@@ -1,88 +1,166 @@
...
 from rest_framework import exceptions
 
+from addons.twofactor.models import UserSettings
+from api.base.exceptions import (UnconfirmedAccountError, UnclaimedAccountError, DeactivatedAccountError,
+                                 MergedAccountError, InvalidAccountError, TwoFactorRequiredError)
 from framework.auth import cas
-from framework.sessions.model import Session
-from framework.auth.core import User, get_user
+from framework.auth.core import get_user
+from osf.models import OSFUser, Session
```
Comment (cslzchen) [inline api/base/authentication/drf.py]: Use Django models and query.

File: api/base/authentication/drf.py [hunk 3]
```diff
@@ -1,88 +1,166 @@
...
 from website import settings
-from api.base.exceptions import UnconfirmedAccountError, DeactivatedAccountError, TwoFactorRequiredError
+
+UNUSABLE_PASSWORD_PREFIX = hashers.UNUSABLE_PASSWORD_PREFIX
```
Comment (icereval) [inline api/base/authentication/drf.py]: use `hashers` below

File: api/base/authentication/drf.py [hunk 4]
```diff
@@ -1,88 +1,166 @@
...
+    """
+    Given a cookie value, return the `Session` object or `None`.
+
+    :param cookie_val: the cookie
+    :return: the `Session` object or None
+    """
+
     session_id = itsdangerous.Signer(settings.SECRET_KEY).unsign(cookie_val)
-    return Session.load(session_id)
+    return Session.objects.filter(_id=session_id).first()
```
Comment (icereval) [inline api/base/authentication/drf.py]: `Session.objects.get(_id=session_id)` (catch the exception if needed, `Session.ObjectDoesNotExist`)

File: api/base/authentication/drf.py [hunk 5]
```diff
@@ -1,88 +1,166 @@
...
+    # user merged
+    if user.is_merged:
+        raise MergedAccountError
+
+    # user not confimred or contributor not claimed
+    if not user.is_confirmed and not user.is_registered and not user.is_claimed:
+        if user.password is None or user.password.startswith(UNUSABLE_PASSWORD_PREFIX):
+            raise UnclaimedAccountError
+        if user.has_usable_password():
+            raise UnconfirmedAccountError
```
Comment (icereval) [inline api/base/authentication/drf.py]: consolidate

File: api/base/authentication/drf.py [hunk 6]
```diff
@@ -1,88 +1,166 @@
...
+    :raises UnconfirmedAccountError
+    :raises UnclaimedAccountError
+    :raises DeactivatedAccountError
+    :raises MergedAccountError
+    :raises InvalidAccountError
+    """
+
+    # active user must be registered, claimed, confirmed, not merged or disabled, and has a usable password
+    if user.is_active:
+        return
```
Comment (icereval) [inline api/base/authentication/drf.py]: :+1:

File: api/base/authentication/drf.py [hunk 7]
```diff
@@ -1,88 +1,166 @@
...
+        """
         cookie_val = request.COOKIES.get(settings.COOKIE_NAME)
         if not cookie_val:
             return None
         session = get_session_from_cookie(cookie_val)
         if not session:
             return None
         user_id = session.data.get('auth_user_id')
-        user = User.load(user_id)
+        user = OSFUser.objects.filter(guids___id=user_id).first()
```
Comment (icereval) [inline api/base/authentication/drf.py]: `OSFUser.objects.get(guids___id=user_id)` (catch the exception if needed, `OSFUser.ObjectDoesNotExist`)

File: api/base/authentication/drf.py [hunk 8]
```diff
@@ -1,88 +1,166 @@
...
+
+        :param user: the user
+        :param request: the request
+        :raises TwoFactorRequiredError
+        :raises AuthenticationFailed
         """
-        user_addon = user.get_addon('twofactor')
-        if user_addon and user_addon.is_confirmed:
+
+        two_fator = UserSettings.objects.filter(owner_id=user.pk).first()
```
Comment (icereval) [inline api/base/authentication/drf.py]: `two_fator` -> `two_factor`

Comment (icereval) [inline api/base/authentication/drf.py]: same `get` orm change, handle exception

File: api/base/authentication/drf.py [hunk 9]
```diff
@@ -1,88 +1,166 @@
 import itsdangerous
+
+from django.contrib.auth import hashers
 from django.utils.translation import ugettext_lazy as _
 
 from rest_framework import authentication
 from rest_framework.authentication import BasicAuthentication
 from rest_framework import exceptions
 
+from addons.twofactor.models import UserSettings
```
Comment (icereval) [inline api/base/authentication/drf.py]: `from addons.twofactor.models import UserSettings` -> `from addons.twofactor.models import UserSettings as TwoFactorUserSettings`

File: api/base/authentication/drf.py [hunk 10]
```diff
@@ -119,12 +197,15 @@ def authenticate(self, request):
         if cas_auth_response.authenticated is False:
             raise exceptions.NotAuthenticated(_('CAS server failed to authenticate this token'))
 
-        user = User.objects.filter(guids___id=cas_auth_response.user).first()
+        user = OSFUser.objects.filter(guids___id=cas_auth_response.user).first()
```
Comment (icereval) [inline api/base/authentication/drf.py]: same `get` orm change, consolidate w/ error handler below

File: api/base/authentication/drf.py [hunk 11]
```diff
@@ -1,88 +1,171 @@
 import itsdangerous
+
 from django.utils.translation import ugettext_lazy as _
 
 from rest_framework import authentication
 from rest_framework.authentication import BasicAuthentication
 from rest_framework import exceptions
 
+from addons.twofactor.models import UserSettings as TwoFactorUserSettings
```
Comment (cslzchen) [inline api/base/authentication/drf.py]: Update for CR: use alias `TwoFactorUserSettings`.

File: api/base/authentication/drf.py [hunk 12]
```diff
@@ -1,88 +1,171 @@
...
+    Given a cookie value, return the `Session` object or `None`.
+
+    :param cookie_val: the cookie
+    :return: the `Session` object or None
+    """
+
     session_id = itsdangerous.Signer(settings.SECRET_KEY).unsign(cookie_val)
-    return Session.load(session_id)
+    try:
+        session = Session.objects.get(_id=session_id)
```
Comment (cslzchen) [inline api/base/authentication/drf.py]: Update for CR: use `.get()` with exception handling.

File: api/base/authentication/drf.py [hunk 13]
```diff
@@ -1,88 +1,171 @@
...
         cookie_val = request.COOKIES.get(settings.COOKIE_NAME)
         if not cookie_val:
             return None
         session = get_session_from_cookie(cookie_val)
         if not session:
             return None
         user_id = session.data.get('auth_user_id')
-        user = User.load(user_id)
+        try:
+            user = OSFUser.objects.get(guids___id=user_id)
```
Comment (cslzchen) [inline api/base/authentication/drf.py]: Update for CR: use .get() with exception handling.

File: api/base/authentication/drf.py [hunk 14]
```diff
@@ -1,88 +1,171 @@
...
+        :param user: the user
+        :param request: the request
+        :raises TwoFactorRequiredError
+        :raises AuthenticationFailed
         """
-        user_addon = user.get_addon('twofactor')
-        if user_addon and user_addon.is_confirmed:
+
+        try:
+            two_factor = TwoFactorUserSettings.objects.get(owner_id=user.pk)
```
Comment (cslzchen) [inline api/base/authentication/drf.py]: Update for CR: use .get() with exception handling.

File: api/base/authentication/drf.py [hunk 15]
```diff
@@ -119,12 +202,18 @@ def authenticate(self, request):
         if cas_auth_response.authenticated is False:
             raise exceptions.NotAuthenticated(_('CAS server failed to authenticate this token'))
 
-        user = User.objects.filter(guids___id=cas_auth_response.user).first()
+        try:
+            user = OSFUser.objects.get(guids___id=cas_auth_response.user)
```
Comment (cslzchen) [inline api/base/authentication/drf.py]: Update for CR: use .get() with exception handling.

File: api/base/authentication/drf.py [hunk 16]
```diff
@@ -1,88 +1,171 @@
...
-    elif not user.is_confirmed:
-        raise UnconfirmedAccountError
+
+    # user merged
+    if user.is_merged:
+        raise MergedAccountError
+
+    # user not confirmed or contributor not claimed
+    if not user.is_confirmed and not user.is_registered and not user.is_claimed:
+        if user.has_usable_password():
```
Comment (cslzchen) [inline api/base/authentication/drf.py]: Update for CR: use `has_usuable_password()`.

**Pull request merged**